jctools-experimental/src/main/java/org/jctools/uc/FlatCombining.java

```diff
@@ -0,0 +1,181 @@
+ * Negative scalability is when the system-wide throughput drops as the number of active threads increases.
+ * <p>
+ * The Flat Combining concept is described by the paper
+ * <a href="https://www.cs.bgu.ac.il/~hendlerd/papers/flat-combining.pdf">
+ * Flat Combining and the Synchronization-Parallelism Tradeoff</a>
+ *
+ * @param <T> The type of data managed by this flat combining instance.
+ */
+public class FlatCombining<T> {
+    private final WFIStack<OpNode> ops;
+    private final AtomicBoolean lock;
+    private final T data;
+
+    public FlatCombining(T sequentialData) {
+        ops = new WFIStack<OpNode>();
+        lock = new AtomicBoolean();
+        data = sequentialData;
+    }
+
+    /**
+     * Apply the given operation on the underlying data structure. This call may block. Multiple threads can calls
+     * this at the same time. The operations will be applied with linearizable consistency.
+     * <p>
+     * The operation will only be applied once to the data structure, but there is no guarantee that it will be
+     * applied by the calling thread.
+     *
+     * @param operation The operation to apply to the underlying data structure.
+     * @param <R>       The type of return value produced by the operation, if any.
+     * @param <E>       The type of exception the operation might throw, if any.
+     * @return The result returned by the operation.
+     * @throws E                    If the operation throws an exception.
+     * @throws InterruptedException If the apply method ends up blocking but the thread is interrupted.
+     */
+    public <R, E extends Exception> R apply(Operation<T, R, E> operation) throws E, InterruptedException {
+        OpNode op = enqueue(operation);
+
+        for (; ; ) {
+            if (tryLock()) {
+                Iterable<OpNode> nodes = combineAndApplyAllOperations();
+                unlockAndUnparkWaiters(nodes);
+            } else {
+                op.await();
+            }
+            if (op.isDone()) {
+                try {
+                    return op.getResult();
+                } catch (InterruptedException ie) {
+                    throw ie;
+                } catch (Exception e) {
+                    //noinspection unchecked
+                    throw (E) e;
+                } catch (Throwable th) {
+                    throw new RuntimeException(th);
+                }
+            }
+        }
+    }
+
+    private <R, E extends Exception> OpNode enqueue(Operation<T, R, E> operation) {
+        if (operation == null) {
+            throw new NullPointerException("The operation cannot be null.");
+        }
+        OpNode op = new OpNode(operation);
+        ops.push(op);
+        return op;
+    }
+
+    private boolean tryLock() {
+        return !lock.get() && lock.compareAndSet(false, true);
+    }
+
+    private Iterable<OpNode> combineAndApplyAllOperations() {
+        Iterable<OpNode> nodes = ops.popAllFifo();
```

Comment:
This is a special single consumer operation, because protected by the lock ie the stack could expose a lighter single consumer (or exclusive consumer) variant of this (or a weakPopAllFifo that can fail under contention too; a condition they we won't expect to happen here ;))

Reply:
You're imagining a `tryApply` method that only applies if it can do so without blocking?

I'm not a fan of that, since we have to enqueue before we can make an attempt at the lock. If getting the lock fails, we'd have cancellation racing with another thread applying our operation. Even if that gets coordinated, we'd still leave our operation enqueued until another thread comes to apply.

Reply:
For instance, we don't know how long an operation is going to take, and once it has been _started_ we can't return from `apply` until it's been completed. Otherwise I think we'll get a programming model that's hard to reason about.

Reply:
You're right :)
I've now looked how the stack is implementing `popAllFifo` and I've forgotten that's not a queue ie it doesn't have separated head and tail pointers, hence cannot implement a `single-writer` variant of poll.
Instad, I see that `MpscLinkedQueue` already implement a single-consumer FIFO (by definition) poll and has a wait-free `enqueue` as well: maybe the `FlatCombining` construct could abstract which `fifo` data-structure to use eg `MpscLinkedQueue`,`WFIStack` or other...